src/mlpack/methods/det/dt_utils.hpp

```diff
@@ -67,7 +67,51 @@ DTree<MatType, TagType>* Trainer(MatType& dataset,
                                  const bool useVolumeReg = false,
                                  const size_t maxLeafSize = 10,
                                  const size_t minLeafSize = 5,
-                                 const std::string unprunedTreeOutput = "");
+                                 const std::string unprunedTreeOutput = "",
+                                 const bool skipPruning = false);
+
+/**
+ * The class responsible for cacheing the path to each node of the tree. Its instance
+ * is provided to EnumerateTree() utility ONCE and it caches the paths to all the
+ * leafs and then easily (and quickly) retrieves these paths for each test entry.
+ */
+class PathCacher
+{
+ public:
+  enum PathFormat
+  {
+    FormatLR,
+    FormatLR_ID,
+    FormatID_LR
+  };
+
+  template <typename MatType>
+  PathCacher(PathFormat fmt, DTree<MatType, int>* tree);
+
+  template <typename MatType>
+  void  Enter(const DTree<MatType, int>* node,
+              const DTree<MatType, int>* parent);
+
+  template <typename MatType>
+  void  Leave(const DTree<MatType, int>* node,
+              const DTree<MatType, int>* parent);
+
+  const std::string&  PathFor(int tag) const;
+
+  int                 ParentOf(int tag) const;
+
+  size_t              NumNodes() const { return pathCache.size(); }
+
+ protected:
+  typedef std::list<std::pair<bool, int> >            PathType;
+  typedef std::vector<std::pair<int, std::string> >   PathCacheType;
+
+  PathType      path;
+  PathFormat    format;
+  PathCacheType pathCache;
+
+  std::string   BuildString();
```

Comment:
Do you mind if I adjust the spacing (i.e. remove extra spaces between types and functions or types and variable names) after merge for consistency with the rest of the code?  Alternately you can do it, but I don't want to waste your time with a million tiny style issues. :)

Reply:
I'm perfectly fine if you update the spacing, of course, but there are more fixes - like the `enumerate_tree.hpp` comment for `TreeType`. If that is ok for you to make all the changes - please, do so.

Regarding the `TagType` - yes, you have good memory - it was me, who introduced it. My idea was to provide flexible mechanism and ability to tag the node with whatever type you provide - just by overloading the incremental operator, you can produce sequential ids of any type. Now, saying this, I realize that `EnumerateTree()` functionality is not that different, but it comes the question about providing enough context (like the parent). Also tree enumeration is now available for other trees as well. 

Probably the static_assert approach is best for now and letting the idea go in either direction when there are more specific tasks that will need/utilize it.